File: v4/rust/iprange-livedb/src/recovery/id_table.rs
//! Shared fixed-record recovery storage and source-ID lookup.

use std::marker::PhantomData;

use crate::contract::{u32_le, u64_le};
use crate::error::{Error, Result};

use super::tables::{
    hash_u32, Layout, Region, Tables, ID_SLOT_SIZE, MEMBERSHIP_RECORD_SIZE, STRUCTURE_RECORD_SIZE,
};

const SLOT_SIZE: usize = ID_SLOT_SIZE as usize;
const OCCUPIED_OFFSET: usize = 0;
const CONFLICT_OFFSET: usize = 1;
const ID_OFFSET: usize = 4;
const RECORD_OFFSET: usize = 8;
const RECORD_BUFFER_SIZE: usize = if MEMBERSHIP_RECORD_SIZE > STRUCTURE_RECORD_SIZE {
    MEMBERSHIP_RECORD_SIZE as usize
} else {
    STRUCTURE_RECORD_SIZE as usize
};

pub(crate) enum Insert {
    First,
    Duplicate { first: u64, newly_conflicted: bool },
}

pub(crate) trait RecordCodec {
    type Record: Copy;

    const WIDTH: usize;
    const INVALID_RECORD: &'static str;
    const FULL: &'static str;

    fn regions(layout: Layout) -> (Region, Region);
    fn encode(record: Self::Record, output: &mut [u8]);
    fn decode(input: &[u8]) -> Result<Self::Record>;
    fn rejected(record: Self::Record) -> bool;
    fn reject(record: &mut Self::Record);
}

pub(crate) struct IdIndex<C> {
    table: RawIdTable,
    marker: PhantomData<C>,
}

impl<C: RecordCodec> IdIndex<C> {
    pub(crate) fn new(tables: &Tables) -> Self {
        let (records, ids) = C::regions(tables.layout());
        debug_assert!(C::WIDTH <= RECORD_BUFFER_SIZE);
        Self {
            table: RawIdTable::new(records, ids, C::INVALID_RECORD, C::FULL),
            marker: PhantomData,
        }
    }

    pub(crate) fn push(&mut self, tables: &mut Tables, record: C::Record) -> Result<()> {
        let mut bytes = [0; RECORD_BUFFER_SIZE];
        C::encode(record, &mut bytes[..C::WIDTH]);
        self.table.push(tables, &bytes[..C::WIDTH])
    }

    pub(crate) fn record(&self, tables: &Tables, index: u64) -> Result<C::Record> {
        let mut bytes = [0; RECORD_BUFFER_SIZE];
        self.table.read(tables, index, &mut bytes[..C::WIDTH])?;
        C::decode(&bytes[..C::WIDTH])
    }

    pub(crate) fn reject(&self, tables: &mut Tables, index: u64) -> Result<()> {
        let mut record = self.record(tables, index)?;
        C::reject(&mut record);
        let mut bytes = [0; RECORD_BUFFER_SIZE];
        C::encode(record, &mut bytes[..C::WIDTH]);
        self.table.write(tables, index, &bytes[..C::WIDTH])
    }

    pub(crate) fn insert_id(
        &self,
        tables: &mut Tables,
        id: u32,
        record_index: u64,
    ) -> Result<Insert> {
        self.table.insert_id(tables, id, record_index)
    }

    pub(crate) fn get(&self, tables: &Tables, id: u32) -> Result<Option<C::Record>> {
        let Some(index) = self.table.get(tables, id)? else {
            return Ok(None);
        };
        let record = self.record(tables, index)?;
        Ok((!C::rejected(record)).then_some(record))
    }

    pub(crate) const fn records_len(&self) -> u64 {
        self.table.records_len()
    }
}

struct RawIdTable {
    records: Region,
    ids: Region,
    records_len: u64,
    invalid_record: &'static str,
    full: &'static str,
}

impl RawIdTable {
    pub(crate) const fn new(
        records: Region,
        ids: Region,
        invalid_record: &'static str,
        full: &'static str,
    ) -> Self {
        Self {
            records,
            ids,
            records_len: 0,
            invalid_record,
            full,
        }
    }

    pub(crate) fn push(&mut self, tables: &mut Tables, record: &[u8]) -> Result<()> {
        if self.records_len == self.records.slots {
            return Err(Error::RecoveryCandidateChanged);
        }
        tables.write(self.records, self.records_len, record)?;
        self.records_len += 1;
        Ok(())
    }

    pub(crate) fn read(&self, tables: &Tables, index: u64, output: &mut [u8]) -> Result<()> {
        self.require_record(index)?;
        tables.read(self.records, index, output)
    }

    pub(crate) fn write(&self, tables: &mut Tables, index: u64, input: &[u8]) -> Result<()> {
        self.require_record(index)?;
        tables.write(self.records, index, input)
    }

    pub(crate) fn insert_id(
        &self,
        tables: &mut Tables,
        id: u32,
        record_index: u64,
    ) -> Result<Insert> {
        self.require_record(record_index)?;
        let (slot_index, mut slot) = match self.probe(tables, id, false)? {
            Some(found) => found,
            None => self
                .probe(tables, id, true)?
                .ok_or(Error::Corrupt(self.full))?,
        };
        if slot[OCCUPIED_OFFSET] == 0 {
            slot[OCCUPIED_OFFSET] = 1;
            slot[ID_OFFSET..RECORD_OFFSET].copy_from_slice(&id.to_le_bytes());
            slot[RECORD_OFFSET..SLOT_SIZE].copy_from_slice(&record_index.to_le_bytes());
            tables.write(self.ids, slot_index, &slot)?;
            return Ok(Insert::First);
        }
        let newly_conflicted = slot[CONFLICT_OFFSET] == 0;
        slot[CONFLICT_OFFSET] = 1;
        tables.write(self.ids, slot_index, &slot)?;
        Ok(Insert::Duplicate {
            first: u64_le(&slot, RECORD_OFFSET),
            newly_conflicted,
        })
    }

    pub(crate) fn get(&self, tables: &Tables, id: u32) -> Result<Option<u64>> {
        let Some((_, slot)) = self.probe(tables, id, false)? else {
            return Ok(None);
        };
        if slot[CONFLICT_OFFSET] != 0 {
            return Ok(None);
        }
        let index = u64_le(&slot, RECORD_OFFSET);
        self.require_record(index)?;
        Ok(Some(index))
    }

    pub(crate) const fn records_len(&self) -> u64 {
        self.records_len
    }

    fn require_record(&self, index: u64) -> Result<()> {
        if index >= self.records_len {
            return Err(Error::Corrupt(self.invalid_record));
        }
        Ok(())
    }

    fn probe(
        &self,
        tables: &Tables,
        id: u32,
        empty: bool,
    ) -> Result<Option<(u64, [u8; SLOT_SIZE])>> {
        if self.ids.slots == 0 {
            return Ok(None);
        }
        let mask = self.ids.slots - 1;
        let mut index = hash_u32(id) & mask;
        for _ in 0..self.ids.slots {
            let mut slot = [0; SLOT_SIZE];
            tables.read(self.ids, index, &mut slot)?;
            if slot[OCCUPIED_OFFSET] == 0 {
                return Ok(empty.then_some((index, slot)));
            }
            if u32_le(&slot, ID_OFFSET) == id {
                return Ok((!empty).then_some((index, slot)));
            }
            index = (index + 1) & mask;
        }
        Ok(None)
    }
}

File: v4/go/internal/recovery/id_table.go
package recovery

// Shared fixed-record recovery storage and source-ID lookup (Rust
// recovery/id_table.rs): the record region and the ID slot region of
// one record class, the push/read/write/reject terminals, and the ID
// probe with the conflict classes.

import "github.com/firehol/iprange/v4/go/internal/format"

const (
	idSlotSize       = uint64(16)
	idSlotOccupied   = 0
	idSlotConflict   = 1
	idSlotID         = 4
	idSlotRecord     = 8
	idSlotRecordSize = 16
)

// idInsert is the ID-slot outcome of one registration (Rust Insert).
type idInsert struct {
	first           uint64 // first record of a conflicting slot
	duplicate       bool
	newlyConflicted bool
}

// recordCodec adapts one record class to the shared id table (Rust
// RecordCodec).
type recordCodec[R any] struct {
	width         int
	invalidRecord string
	full          string
	regions       func(layout tableLayout) (records, ids tableRegion)
	encode        func(record R, output []byte)
	decode        func(input []byte) (R, error)
	isRejected    func(record R) bool
	reject        func(record *R)
}

// idIndex is the shared record and ID-slot index of one recovery table
// (Rust IdIndex).
type idIndex[R any] struct {
	codec recordCodec[R]
	table rawIDTable
}

// newIDIndex builds one index over the store layout (Rust IdIndex::new).
func newIDIndex[R any](tables *tableStore, codec recordCodec[R]) *idIndex[R] {
	records, ids := codec.regions(tables.layout)
	return &idIndex[R]{codec: codec, table: rawIDTable{records: records, ids: ids, invalidRecord: codec.invalidRecord, full: codec.full}}
}

// push appends one record (Rust IdIndex::push: the record bound is the
// candidate-changed class).
func (c *idIndex[R]) push(tables *tableStore, record R) error {
	bytes := make([]byte, c.codec.width)
	c.codec.encode(record, bytes)
	return c.table.push(tables, bytes)
}

// record reads and decodes one record (Rust IdIndex::record).
func (c *idIndex[R]) record(tables *tableStore, index uint64) (R, error) {
	var out R
	bytes := make([]byte, c.codec.width)
	if err := c.table.read(tables, index, bytes); err != nil {
		return out, err
	}
	return c.codec.decode(bytes)
}

// reject marks one record rejected (Rust IdIndex::reject).
func (c *idIndex[R]) reject(tables *tableStore, index uint64) error {
	record, err := c.record(tables, index)
	if err != nil {
		return err
	}
	c.codec.reject(&record)
	bytes := make([]byte, c.codec.width)
	c.codec.encode(record, bytes)
	return c.table.write(tables, index, bytes)
}

// insertID registers one record under its ID (Rust IdIndex::insert_id).
func (c *idIndex[R]) insertID(tables *tableStore, id uint32, recordIndex uint64) (idInsert, error) {
	return c.table.insertID(tables, id, recordIndex)
}

// get returns the un-rejected record of one ID (Rust IdIndex::get).
func (c *idIndex[R]) get(tables *tableStore, id uint32) (R, bool, error) {
	var out R
	index, found, err := c.table.get(tables, id)
	if err != nil || !found {
		return out, false, err
	}
	record, err := c.record(tables, index)
	if err != nil {
		return out, false, err
	}
	return record, !c.codec.isRejected(record), nil
}

// recordsLen returns the record count.
func (c *idIndex[R]) recordsLen() uint64 {
	return c.table.recordsLen
}

// rawIDTable is the region-level record and ID table (Rust
// RawIdTable).
type rawIDTable struct {
	records       tableRegion
	ids           tableRegion
	recordsLen    uint64
	invalidRecord string
	full          string
}

// push appends one encoded record.
func (t *rawIDTable) push(tables *tableStore, record []byte) error {
	if t.recordsLen == t.records.slots {
		return candidateChangedError()
	}
	if err := tables.write(t.records, t.recordsLen, record); err != nil {
		return err
	}
	t.recordsLen++
	return nil
}

// read reads one record with the record-index proof (Rust RawIdTable::
// read).
func (t *rawIDTable) read(tables *tableStore, index uint64, output []byte) error {
	if err := t.requireRecord(index); err != nil {
		return err
	}
	return tables.read(t.records, index, output)
}

// write rewrites one record.
func (t *rawIDTable) write(tables *tableStore, index uint64, input []byte) error {
	if err := t.requireRecord(index); err != nil {
		return err
	}
	return tables.write(t.records, index, input)
}

// insertID registers one ID slot (Rust RawIdTable::insert_id: the
// first registration, or the duplicate with the first-record proof and
// the new-conflict envelope fact).
func (t *rawIDTable) insertID(tables *tableStore, id uint32, recordIndex uint64) (idInsert, error) {
	if err := t.requireRecord(recordIndex); err != nil {
		return idInsert{}, err
	}
	slotIndex, slot, found, err := t.probe(tables, id, false)
	if err != nil {
		return idInsert{}, err
	}
	if !found {
		slotIndex, slot, found, err = t.probe(tables, id, true)
		if err != nil {
			return idInsert{}, err
		}
		if !found {
			return idInsert{}, corruptError(t.full)
		}
	}
	if slot[idSlotOccupied] == 0 {
		slot[idSlotOccupied] = 1
		format.PutU32(slot[idSlotID:idSlotRecord], id)
		format.PutU64(slot[idSlotRecord:idSlotRecordSize], recordIndex)
		if err := tables.write(t.ids, slotIndex, slot[:]); err != nil {
			return idInsert{}, err
		}
		return idInsert{}, nil
	}
	newlyConflicted := slot[idSlotConflict] == 0
	slot[idSlotConflict] = 1
	if err := tables.write(t.ids, slotIndex, slot[:]); err != nil {
		return idInsert{}, err
	}
	return idInsert{
		duplicate:       true,
		first:           format.U64(slot[idSlotRecord:idSlotRecordSize]),
		newlyConflicted: newlyConflicted,
	}, nil
}

// get resolves one un-conflicted ID to its record index (Rust
// RawIdTable::get).
func (t *rawIDTable) get(tables *tableStore, id uint32) (uint64, bool, error) {
	_, slot, found, err := t.probe(tables, id, false)
	if err != nil || !found {
		return 0, false, err
	}
	if slot[idSlotConflict] != 0 {
		return 0, false, nil
	}
	index := format.U64(slot[idSlotRecord:idSlotRecordSize])
	if err := t.requireRecord(index); err != nil {
		return 0, false, err
	}
	return index, true, nil
}

// requireRecord proves the record index (Rust require_record: the
// Corrupt class).
func (t *rawIDTable) requireRecord(index uint64) error {
	if index >= t.recordsLen {
		return corruptError(t.invalidRecord)
	}
	return nil
}

// probe probes one ID slot (Rust RawIdTable::probe: the
// occupied-or-match acceptance; empty probes return the empty slot).
func (t *rawIDTable) probe(tables *tableStore, id uint32, empty bool) (uint64, [idSlotRecordSize]byte, bool, error) {
	var slot [idSlotRecordSize]byte
	if t.ids.slots == 0 {
		return 0, slot, false, nil
	}
	mask := t.ids.slots - 1
	index := hashU32(id) & mask
	for probe := uint64(0); probe < t.ids.slots; probe++ {
		if err := tables.read(t.ids, index, slot[:]); err != nil {
			return 0, slot, false, err
		}
		if slot[idSlotOccupied] == 0 {
			if empty {
				return index, slot, true, nil
			}
			return 0, slot, false, nil
		}
		if format.U32(slot[idSlotID:idSlotRecord]) == id {
			if !empty {
				return index, slot, true, nil
			}
			return 0, slot, false, nil
		}
		index = (index + 1) & mask
	}
	return 0, slot, false, nil
}
